File: Data Structures & Algorithms/sort-colors/submission-3.swift
class Solution {
    func sortColors(_ nums: inout [Int]) {
        
        var left = 0
        var right = nums.count - 1
        var i = 0

        while i <= right {
            if nums[i] == 0 {
                swap(&nums, left, i)
                left += 1
                i += 1
            } else if nums[i] == 2 {
                swap(&nums, right, i)
                right -= 1
            } else {
                i += 1
            }
        }

    }

    func swap(_ nums: inout [Int], _ i: Int, _ j: Int) {
        let temp = nums[i]
        nums[i] = nums[j]
        nums[j] = temp
    }
}

File: Data Structures & Algorithms/house-robber/submission-3.kt
class Solution {
    fun rob(nums: IntArray): Int {

        var cache = HashMap<Int, Int>()

        fun dfs(i: Int): Int {
            if (i >= nums.count()) {
                return 0
            }
            if (cache[i] != null) {
                return cache[i]!!
            }

            // rob at i
            var money1 = nums[i] + dfs(i + 2)
            // skip at i
            var money2 = dfs(i + 1)

            cache[i] = maxOf(money1, money2)

            return cache[i]!!
        }

        return dfs(0)

        // if (nums.count() < 0) {
        //     return -1
        // }

        // if (nums.count() == 1) {
        //     return nums[0]
        // }


        // var house1 = nums[0]
        // var house2 = maxOf(house1, nums[1])

        // for (i in 2 ..< nums.count()) {
        //     val temp = maxOf(nums[i] + house1, house2)
        //     house1 = house2
        //     house2 = temp
        // }

        // return house2

    }
}
